spec/index.md

```diff
@@ -1652,6 +1654,26 @@ There must be at most one call to `ic0.call_on_cleanup` between `ic0.call_new` a
 
     This system call traps if the cycle balance of the canister after transferring cycles decreases below the canister's freezing limit.
 
+-   `ic0.call_cycles_add128_up_to : (max_amount_high : i64, max_amount_low : i64, dst: i32) -> ()`
+
+    This system call moves cycles from the canister balance onto the call under construction, to be transferred with that call.
+
+    This moves the maximum possible amount of cycles onto the call, up to these constraints:
+
+    - It moves no more cycles than represented by a 128-bit value which can be obtained by combining the `max_amount_high` and `max_amount_low` parameters.
+
+    - A subsequent `ic0.call_perform` does not fail because of insufficient cycles balance
+      (if no extra bytes were added to the message via `ic0.call_data_append` between `ic0.call_cycles_add128_up_to` and `ic0.call_perform`)
```

Comment:
```suggestion
      (if the canister makes no memory allocations and no extra bytes were added to the message via `ic0.call_data_append` between `ic0.call_cycles_add128_up_to` and `ic0.call_perform`)
```

Reply:
@dsarlis this seems a bit unpredictable to use. In higher-level languages, it seems difficult to guarantee that no extra memory is allocated. Should we maybe extend the implementation so that we can drop the additional conditions (by doing the computation of actual cycles to be sent to `call_perform`)?

Reply:
@mraszyk just pointed out that this would violate
> This system call also copies the actual amount of cycles that were moved onto the call represented [...]
which seems to make it necessary to eagerly compute the amount of cycles attached. Not sure which is the best path forward here.

Reply:
It is possible to compute the actual amount of cycles attached by taking the difference of the canister's cycles balance before and after calling the system API `ic0.call_cycles_add128_up_to` (confirmed empirically). So dropping the destination pointer should not make the API substantially harder to use. The "trick" with computing the cycles balance difference does not work across `ic0.call_perform` though (as xnet fees are charged, too). Moreover, it is not clear what the semantics of multiple `ic0.call_cycles_add128_up_to` should be if we only attach the cycles during `ic0.call_perform` (the sum vs the maximum?). Hence, delaying the actual attachment to `ic0.call_perform` might not be the best path forward either.

I'd suggest to simply drop the destination pointer if the CDK teams could confirm that allocations are unlikely in between `ic0.call_cycles_add128_up_to` and `ic0.call_perform`. (Allocations happening after `ic0.call_perform` are fine as long as they are covered by the unused execution cycles that are refunded first.)